File: 0001-0500/0155.Min Stack.rb
class MinStack

=begin
    initialize your data structure here.
=end
    def initialize()
        @stack = []
    end


=begin
    :type x: Integer
    :rtype: Void
=end
    def push(x)
        mn = @stack.empty? ? x : [x, @stack[-1][-1]].min
        @stack << [x, mn]
    end


=begin
    :rtype: Void
=end
    def pop()
        @stack.pop[-1]
    end


=begin
    :rtype: Integer
=end
    def top()
        @stack[-1][0]
    end


=begin
    :rtype: Integer
=end
    def get_min()
        @stack[-1][-1]
    end


end

# Your MinStack object will be instantiated and called as such:
# obj = MinStack.new()
# obj.push(x)
# obj.pop()
# param_3 = obj.top()
# param_4 = obj.get_min()

File: 0001-0500/0155.Min Stack.java
/**
 * Basic idea:
 * 	deep inside stack is invisible, current info is mentained at stack top.
 * Result:
 * 	17 / 17 test cases passed.
 * 	Status: Accepted
 * 	Runtime: 8 ms
 * 	Your runtime beats 73.97% of java submissions.
 * Date:
 * 	/2/7/2016
 */
import java.util.LinkedList;
class MinStack {
    private class Elem{
        Elem(int val,int min){
            this.val=val;
            this.min = min;
        }
        Elem(int val){
            this.val = val;
            this.min = val;
        }
        int val;
        int min;
    }
    private LinkedList<Elem> stack = new LinkedList<Elem>();
    public void push(int x) {
        if(stack.size()==0)
            stack.add(new Elem(x));
        else
            stack.add(new Elem(x,Math.min(x,stack.getLast().min)));
    }

    public void pop() {
        stack.removeLast();
    }

    public int top() {
        return stack.getLast().val;
    }

    public int getMin() {
        return stack.getLast().min;
    }
}

